### PR TITLE
[DO NOT MRG] plot_compare_evoked improved line colouring

### PR DESCRIPTION
**Pull request by jona-sassenhagen**
Redo of #4221

Good for parametric and complex factorial designs.

See [gist](http://nbviewer.jupyter.org/gist/jona-sassenhagen/95b450e8e4cf558c5c80d026f7c194f0/sequential_colors_demo_2.ipynb)

### CODE REVIEW

**Comment by jona-sassenhagen** [comment]
@mmagnuski can you eventually check if the API works for you? I remember your ICON poster

**Comment by agramfort** [comment]
I propose to complete this once we have metadata in + kword file in datasets. This gist will make an amazing tutorial ! really nice @jona-sassenhagen 👍

**Comment by jona-sassenhagen** [comment]
Yeah, makes sense.

We can in fact reproduce the full analysis in Dufau, Grainger, ... 2015 in a few lines of code with MNE. I'd like to demonstrate how to use the cluster correction, cause it seems to confuse many newcomers, but is very hip in ERPland these days.

**Comment by agramfort** [comment]
@jona-sassenhagen can you rebase?

what's left to do here?

**Comment by jona-sassenhagen** [comment]
Well you said: 

>I propose to complete this once we have metadata in + kword file in datasets

?

**Comment by jona-sassenhagen** [comment]
Or do you just mean the tutorial.

**Comment by agramfort** [comment]
my bad.... forget it...

**Comment by larsoner** [comment]
Yeah this is highly related to / mutually dependent on #4414, and we don't want that in before release. So we'll have to wait. But after we release, as far as I'm concerned this can go in first then we can simplify #4414

**Comment by jona-sassenhagen** [comment]
Eric, try the plot_metadata_epochs tutorial with this :)

**Comment by larsoner** [comment]
Can you rebase and touch the tutorial to make circle do the work?

**Comment by jona-sassenhagen** [comment]
I just did both - is there anything wrong?

**Comment by jona-sassenhagen** [comment]
Oh god, there are tons of errors.

**Comment by larsoner** [comment]
You need to make some change (even a trivial rewrite / typo fix) `tutorials/plot_epochs_metadata.py` otherwise Circle won't re-render it until merge, see:

https://6222-1301584-gh.circle-artifacts.com/0/tmp/circle-artifacts.1VuPluS/html/auto_tutorials/plot_metadata_epochs.html

**Comment by jona-sassenhagen** [comment]
Oh I see, I had those changes locally, but they were not pushed after the rebase (I assume because they were not changed by the original PR).

**Comment by larsoner** [comment]
Nice!

https://6223-1301584-gh.circle-artifacts.com/0/tmp/circle-artifacts.feoSvsz/html/auto_tutorials/plot_metadata_epochs.html

I would separate the last two plots into separate sections so they appear full-size

**Comment by jona-sassenhagen** [comment]
Ugh, these look terrible :D They're supposed to look like this:

unknown-29

unknown-33

**Comment by larsoner** [comment]
CircleCI doesn't have `viridis`, its `matplotlib` is too old

**Comment by jona-sassenhagen** [comment]
Oh wow, we are green, but I need to rebase from my own 2-line PR.

:|

**Comment by agramfort** [comment]
I am sure you will survive...

**Comment by larsoner** [comment]
@jona-sassenhagen once it comes back can you check CircleCI and post a link?

**Comment by jona-sassenhagen** [comment]
I will try :D

**Comment by jona-sassenhagen** [comment]
Hm. [Something is wrong here.](https://6250-1301584-gh.circle-artifacts.com/0/tmp/circle-artifacts.rjZgxRY/html/auto_tutorials/plot_metadata_epochs.html)

**Comment by jona-sassenhagen** [comment]
I get the same for my local Python 2.7 :(

**Comment by larsoner** [comment]
Time to work on those debugging skills

**Comment by jona-sassenhagen** [comment]
Well - better.
https://6255-1301584-gh.circle-artifacts.com/0/tmp/circle-artifacts.dIS8NwY/html/auto_tutorials/plot_metadata_epochs.html

**Comment by agramfort** [comment]
indeed

ready to go?

**Comment by jona-sassenhagen** [comment]
I would prefer to do a few more tests. With my last big PR, I let too many bugs creep in.

**Comment by larsoner** [comment]
PEP8:

https://travis-ci.org/mne-tools/mne-python/jobs/290636519

**Comment by jona-sassenhagen** [comment]
Ok, I think this is ready for reviews.

**Comment by jona-sassenhagen** [comment]
@mmagnuski when you have time for MNE, maybe also give this one a try ...

**Comment by agramfort** [comment]
@agramfort something changed the post as I was editing and I lost all my comments, so I'm putting it here ...


Factorial:

```
evokeds = {'A/C': evoked1, 'B/C': evoked2, 'A/C': evoked3, 'B/D': evoked4}
mne.viz.plot_compare_evokeds(evokeds, colors="viridis")
or
mne.viz.plot_compare_evokeds(evokeds, colors={"A":"r", "B":"b"}, linestyles={"C":"-", "D":":"})
```

Continuous

```
evokeds = [evoked1, evoked2, evoked3, evoked4]
colors = list(range(len(evokeds)))
cmap = {"colorbar_label":"viridis"}
mne.viz.plot_compare_evokeds(evokeds, colors=colors, cmap=cmap)
```
`colors` is of the same shape as `evokeds`

Mixed:
```
evokeds = {'A/1': evoked1, 'A/2': evoked2, 'B/1': evoked3, 'B/2': evoked4}
linestyles = {"A":"-", "B":":"}
colors = {"1":1, "2":2}
cmap = {"colorbar_label":"jet"}
mne.viz.plot_compare_evokeds(evokeds, colors=colors, linestyles=linestyles, cmap=cmap)
```
(Mixed would depend on 1. giving a `cmap` parameter, 2. `colors` being a dict that links from HED tags to numeric.)

Mixed by order:
**edit:** this actually won't work because we don't have condition names
```
evokeds = [evoked1, evoked2, evoked3, evoked4]
linestyles = ["-", ":", "-", ":"]
colors = [1, 1, 2, 2]
cmap = {"colorbar_label":"jet"}
mne.viz.plot_compare_evokeds(evokeds, colors=colors, linestyles=linestyles, cmap=cmap)

```

(Currently, evokeds can be either 1. an Evoked obj, 2. a list of evoked objs, 3. a dict of evoked objs, 4. a list of lists of evoked objs, 5. a dict of lists of evoked objs.)

Right now I'm not sure the "dict of dicts" approach we discussed would be such a good idea.
`colors = {"condition_label":[{"1":1}, {"2":2}, {"3":3}]}` doesn't look so good.
Creating a pandas series might help, but isn't it overkill when we could simply have a dict?

I think your 'meta' idea would be better for something where we allow more flexibility (e.g., seaborn which has hue, col, row, marker, ...). But here we only ever have two: color and linestyle. Right?

**Comment by agramfort** [comment]
@jona-sassenhagen I just posted what you wrote in the other PR. Let's reuse this for a narrative doc. Don't hesitate to edit.

I see for example that the cmap option is pretty "multi-form" and in the tutorial you only use cmap as a string. Sorry to bug you but I am also trying to teach myself how to use this function.

thanks

**Comment by mmagnuski** [comment]
@jona-sassenhagen I will! :)

**Comment by jona-sassenhagen** [comment]
@agramfort what are you thinking of wrt. a "narrative doc"? Do you mean something like, a tutorial `tutorials/plot_evoked_multiple_conditions.py`?

**Comment by jona-sassenhagen** [comment]
@mmagnuski ..? :)

File: tutorials/plot_metadata_epochs.py [hunk 1]
```diff
@@ -69,35 +69,74 @@
 # of each epoch. We'll create a new column in our metadata object and use
 # it to generate averages for many subsets of trials.
 
-# Create a new metadata column
+# Create two new metadata columns
 meta = epochs.metadata
```
Comment (agramfort) [inline tutorials/plot_metadata_epochs.py]: call it metadata ie replace meta by metadata

File: tutorials/plot_metadata_epochs.py [hunk 2]
```diff
@@ -69,35 +69,74 @@
...
+# For the actual visualisation, we store a number of shared parameters.
 style_plot = dict(
-    colors=plt.get_cmap('YlGnBu_r')(colors),
-    linestyles={'Concrete': '-', 'Abstract': '--'}
+    colors={"Long": "Crimson", "Short": "Cornflowerblue"},
+    linestyles={"Concrete": "-", "Abstract": ":"},
+    split_legend=True,
+    ci=.68,
+    show_sensors=4,
+    show_legend=3,
```
Comment (agramfort) [inline tutorials/plot_metadata_epochs.py]: show_sensors -> sensors_loc
show_legend -> legend_loc

so I understand what these numbers are. Or at least I think i know. It would be even nicer to have:

legend_loc = 'lower left'
sensors_loc = 'upper right'

this should work.

File: tutorials/plot_metadata_epochs.py [hunk 3]
```diff
@@ -69,35 +69,74 @@
...
+plt.show()
+
+###############################################################################
+# To compare words which are 4, 5, 6, 7 or 8 letters long:
+
+evokeds = dict()
+for nlet in epochs.metadata["NumberOfLetters"].unique():
+    evokeds[str(nlet)] = epochs["NumberOfLetters == " + str(nlet)].average()
+
+style_plot["colors"] = {str(nlet): int(nlet) for nlet in
```
Comment (agramfort) [inline tutorials/plot_metadata_epochs.py]: nlet -> n_letters

File: tutorials/plot_metadata_epochs.py (hunk 2 above)
Comment (jona-sassenhagen) [reply tutorials/plot_metadata_epochs.py]: Yes, should have done it like this in the first place. The position numbers for MPL are weird.

File: tutorials/plot_metadata_epochs.py [hunk 4]
```diff
@@ -69,35 +69,74 @@
...
+                        epochs.metadata["NumberOfLetters"].unique()}
+style_plot["cmap"] = "summer_r"
```
Comment (agramfort) [inline tutorials/plot_metadata_epochs.py]: style_plot["cmap"] = ("Number of letters", "summer_r")

File: tutorials/plot_metadata_epochs.py (hunk 2 above)
Comment (agramfort) [reply tutorials/plot_metadata_epochs.py]: forget this for here. I see it has already been released.

Comment (jona-sassenhagen) [reply tutorials/plot_metadata_epochs.py]: Don't forget by default it's simply a boolean switch for showing them or not showing them, so `sensors_loc` probably isn't intuitive.

**Comment by mmagnuski** [comment]
@jona-sassenhagen Hey, sorry, I'll take a look this evening.

**Comment by agramfort** [comment]
@jona-sassenhagen I pushed directly my edits.

please update what's new and you have my +1

**Comment by jona-sassenhagen** [comment]
Thanks!

There is currently a major bug hiding here (I know what it is, just didn't have time to fix) and I'd like @mmagnuski 's input. I'll also take care of sensor locs. Then we can merge.

**Comment by agramfort** [comment]
ok then I put the PR back to WIP :)

**Comment by mmagnuski** [comment]
Hey @jona-sassenhagen, sorry I keep pushing it back, I'm chased by multiple deadlines. I should be able to look into it tomorrow or monday.

File: mne/viz/evoked.py [hunk 5]
```diff
@@ -1483,6 +1484,9 @@ def plot_compare_evokeds(evokeds, picks=list(), gfp=False, colors=None,
         "Vis/L", "Vis/R", `colors` can be `dict(Aud='r', Vis='b')` to map both
         Aud/L and Aud/R to the color red and both Visual conditions to blue.
         If None (default), a sequence of desaturated colors is used.
+        If `cmap` is None, `colors` will indicate how each condition will be
```
Comment (mmagnuski) [inline mne/viz/evoked.py]: `indicate how each condition is colored`

File: mne/viz/evoked.py [hunk 6]
```diff
@@ -1483,6 +1484,9 @@ def plot_compare_evokeds(evokeds, picks=list(), gfp=False, colors=None,
...
+        colored with reference to its position on the colormap - see `colormap`
```
Comment (mmagnuski) [inline mne/viz/evoked.py]: should be: `cmap`

File: mne/viz/evoked.py [hunk 7]
```diff
@@ -1498,9 +1502,31 @@ def plot_compare_evokeds(evokeds, picks=list(), gfp=False, colors=None,
...
         the linewidth for "Aud/L" to 1. Note that HED ('/'-separated) tags are
         not supported.
+    cmap : None | str | tuple
+        If not None, plot evoked activity with colors from a color gradient
+        (indicated by a str referencing a matplotlib colormap - e.g., "viridis"
+        or "Reds").
+        If ``evokeds`` is a list and ``colors`` is `None`, the color will
+        depend on the list position. If ``colors`` is a list, it must contain
+        integers where the list positions correspond to ``evokeds``, and the
+        value corresponds to the position on the colorbar.
```
Comment (mmagnuski) [inline mne/viz/evoked.py]: I am not sure this complexity is needed. If you want a different order of colors, you can reorder Evokeds, and when you need specific colors you can get these from the colormap yourself and pass to `colors`.

File: mne/viz/evoked.py [hunk 8]
```diff
@@ -1498,9 +1502,31 @@ def plot_compare_evokeds(evokeds, picks=list(), gfp=False, colors=None,
...
+        If ``evokeds`` is a dict, ``colors`` should be a dict mapping from
+        (potentially HED-style) condition tags to numbers corresponding to
+        rank order positions on the colorbar. E.g., ::
```
Comment (mmagnuski) [inline mne/viz/evoked.py]: aparat from the remark above - this is documentation for `cmap`, but it is mostly devoted to `colors` kwarg - maybe that should be there?

File: mne/viz/evoked.py [hunk 9]
```diff
@@ -1705,13 +1741,68 @@ def plot_compare_evokeds(evokeds, picks=list(), gfp=False, colors=None,
...
+                                 "``colors`` need to be numeric. Got " +
+                                 str(type(color_value)))
+        cmapper = getattr(plt.cm, cmap, cmap)
+        color_conds = list(colors.keys())
+        all_colors = [colors[cond] for cond in color_conds]
+        n_colors = len(all_colors)
+        color_order = np.array(all_colors).argsort()
+        color_indices = color_order.argsort()
+
+        the_colors = cmapper(np.linspace(0, 1, n_colors))
```
Comment (mmagnuski) [inline mne/viz/evoked.py]: maybe it would be better to ask users to give colors as 0 - 1 floats, then non-linear mapping would be possible (for example only picking values [0.3, 0.5, 0.8])? But I am still not sure if this is better than getting the colors and passing them yourself.

File: mne/viz/evoked.py [hunk 10]
```diff
@@ -1705,13 +1741,68 @@ def plot_compare_evokeds(evokeds, picks=list(), gfp=False, colors=None,
...
+        colors = dict()
+        for cond in evokeds.keys():
+            for cond_number, color in colors_.items():
+                if cond_number in cond:
+                    colors[cond] = the_colors[color]
+                    continue
+
     if not isinstance(colors, dict):  # default colors from M Waskom's Seaborn
         # XXX should put a good list of default colors into defaults.py
         colors_ = ['#e41a1c', '#377eb8', '#4daf4a', '#984ea3', '#ff7f00',
```
Comment (mmagnuski) [inline mne/viz/evoked.py]: just BTW - I am not sure about that, if one is using some matplotlib style maybe we should honor that?

File: mne/viz/evoked.py [hunk 11]
```diff
@@ -1819,18 +1912,39 @@ def plot_compare_evokeds(evokeds, picks=list(), gfp=False, colors=None,
             head_pos = {'center': (0, 0), 'scale': (0.5, 0.5)}
             pos, outlines = _check_outlines(pos, np.array([1, 1]), head_pos)
             if not isinstance(show_sensors, (np.int, bool)):
-                raise TypeError("`show_sensors` must be numeric or bool, not" +
-                                str(type(show_sensors)))
+                raise ValueError("show_sensors must be numeric or bool, not" +
```
Comment (mmagnuski) [inline mne/viz/evoked.py]: isn't `TypeError` more suitable here?

File: mne/viz/evoked.py [hunk 12]
```diff
@@ -1819,18 +1912,39 @@ def plot_compare_evokeds(evokeds, picks=list(), gfp=False, colors=None,
...
+        divider = make_axes_locatable(axes)
+        ax_cb = divider.append_axes("right", size="5%", pad=0.05)
+        ax_cb.imshow(the_colors[:, np.newaxis, :], interpolation='none')
+        ax_cb.set_yticks(np.arange(len(the_colors)))
+        ax_cb.set_yticklabels(np.array(color_conds)[color_order])
+        ax_cb.yaxis.tick_right()
+        ax_cb.set_xticks(())
+        ax_cb.set_ylabel(cmap_label)
+        fig.cbar = ax_cb
+    fig.ts_ax = axes
```
Comment (mmagnuski) [inline mne/viz/evoked.py]: Oh, that's nice, I wasn't aware you can add custom properties to Figure.

File: mne/viz/evoked.py (hunk 10 above)
Comment (jona-sassenhagen) [reply mne/viz/evoked.py]: Well, we need default colors listed explicitly somewhere anyways. That's the way the code works.

File: mne/viz/evoked.py (hunk 11 above)
Comment (jona-sassenhagen) [reply mne/viz/evoked.py]: I always get that confused. But I think this is a ValueError because it's the value of a kwarg.

File: mne/viz/evoked.py (hunk 12 above)
Comment (jona-sassenhagen) [reply mne/viz/evoked.py]: Errrr. That's just monkey patching stuff I think. It's not supported at all, I'm just sticking it on there.

File: mne/viz/evoked.py (hunk 8 above)
Comment (jona-sassenhagen) [reply mne/viz/evoked.py]: I see what you mean. However, here's my rationale. `colors` has essentially two modes: a categorical mode, and a gradient mode. The gradient mode depends on `camp`.

File: mne/viz/evoked.py (hunk 7 above)
Comment (jona-sassenhagen) [reply mne/viz/evoked.py]: Reordering Evokeds won't work for the dict case (which is my preferred case), and won't work for the important case where multiple evokeds get the same color.

Taking colors from an MPL (or, easier, seaborn) colormap is nice and all, but getting *everything*. set up - particularly the split legends - is hell on the user.

**Comment by larsoner** [comment]
Can you incorporate the fixes from #4529 here? If so please close that PR.

Is it easy/possible to also fix #4730 here? If not, let's wait until this is merged to tackle that one so we don't get merge conflicts.

File: mne/viz/evoked.py (hunk 12 above)
Comment (larsoner) [reply mne/viz/evoked.py]: It's better not to monkey patch unless you have a good reason / really need to. I suspect your colorbar axes can still be accessed via `figure.axes` somewhere so this doesn't seem worth it

File: mne/viz/evoked.py (hunk 11 above)
Comment (larsoner) [reply mne/viz/evoked.py]: I don't think Type/Value usually has to do with whether it was positional or keyword, but rather the type of error (so in this case I agree `TypeError` seems more suitable)

File: mne/viz/evoked.py (hunk 10 above)
Comment (larsoner) [reply mne/viz/evoked.py]: technically you could probably extract the default `matplotlib` color cycle, but I don't mind customizing it here (we do it elsewhere, too, I think)

**Comment by agramfort** [comment]
having both colors and cmap was indeed also a question for me...
@jona-sassenhagen
what's your bandwidth to finish this?

File: mne/viz/evoked.py (hunk 8 above)
Comment (mmagnuski) [reply mne/viz/evoked.py]: Yes, I get that, it is just a bit confusing, no big deal.

File: mne/viz/evoked.py (hunk 7 above)
Comment (mmagnuski) [reply mne/viz/evoked.py]: Yes, after looking at how easy and elegant the example some lines below is - I tend to agree.
However as I mentioned in another comment it would be more flexible if colors could get floats between 0 and 1 and the linear order was not assumed.

File: tutorials/plot_metadata_epochs.py (hunk 2 above)
Comment (mmagnuski) [reply tutorials/plot_metadata_epochs.py]: Maybe if we change kwarg to `legend` (or something like that) both `False` and `lower_left` would make sense.
But that was added in previous version, right?

**Comment by mmagnuski** [comment]
The example: [link](https://6346-1301584-gh.circle-artifacts.com/0/tmp/circle-artifacts.2KRrRAL/html/auto_tutorials/plot_metadata_epochs.html).

File: tutorials/plot_metadata_epochs.py [hunk 13]
```diff
@@ -69,36 +71,76 @@
...
 ###############################################################################
+# To compare words which are 4, 5, 6, 7 or 8 letters long:
+
+evokeds = dict()
+for n_letters in epochs.metadata["NumberOfLetters"].unique():
+    n_letters = str(n_letters)
+    evokeds[n_letters] = epochs["NumberOfLetters == " + n_letters].average()
+
+style_plot["colors"] = {str(n_letters): int(n_letters) for n_letters in
+                        epochs.metadata["NumberOfLetters"].unique()}
```
Comment (mmagnuski) [inline tutorials/plot_metadata_epochs.py]: Although I understand the generality, this might be unnecessary noise. It would be more simple to just write:
```
# map keys of evokeds dict to color values
style_plot["colors"] = {'4': 4, '5': 5, '6': 6, '7': 7, '8': 8}
```
(and using floats would be lovely :) )

File: tutorials/plot_metadata_epochs.py [hunk 14]
```diff
@@ -69,36 +71,76 @@
...
-ax.legend(loc=[1.05, .1])
+fig, ax = plt.subplots(figsize=(6, 4))
+mne.viz.plot_compare_evokeds(evokeds, axes=ax, **style_plot)
 plt.show()
 
 ###############################################################################
+# To compare words which are 4, 5, 6, 7 or 8 letters long:
+
+evokeds = dict()
+for n_letters in epochs.metadata["NumberOfLetters"].unique():
```
Comment (mmagnuski) [inline tutorials/plot_metadata_epochs.py]: might be easier to follow with:
```
for n_letters in [4, 5, 6, 7, 8]:
```
although I understand that this demonstrates how to use the metadata in various situations.

File: tutorials/plot_metadata_epochs.py [hunk 15]
```diff
@@ -69,36 +71,76 @@
...
+    n_letters = str(n_letters)
+    evokeds[n_letters] = epochs["NumberOfLetters == " + n_letters].average()
```
Comment (mmagnuski) [inline tutorials/plot_metadata_epochs.py]: would you opt for: `evokeds[str(n_letters)] = ...` instead of two lines? :)

File: tutorials/plot_metadata_epochs.py [hunk 16]
```diff
@@ -69,36 +71,76 @@
...
+mne.viz.plot_compare_evokeds(evokeds, axes=ax, **style_plot)
+plt.show()
+
+###############################################################################
+# And finally, for the interaction between concreteness and continuous length
+# in letters:
+evokeds = dict()
+query = "is_concrete == '{0}' & NumberOfLetters == {1}"
+for concreteness in ("Concrete", "Abstract"):
+    for n_letters in epochs.metadata["NumberOfLetters"].unique():
```
Comment (mmagnuski) [inline tutorials/plot_metadata_epochs.py]: `epochs.metadata["NumberOfLetters"].unique()` appears a number of times in the example, might be worth creating a variable for that - `num_letters_unique` or somthing like that.

File: mne/viz/evoked.py (hunk 12 above)
Comment (jona-sassenhagen) [reply mne/viz/evoked.py]: I mean, it's not exposed. It's just for making it easier internally.

I would prefer to avoid figure.axes cause we have a varying number of axes.

File: mne/viz/evoked.py (hunk 10 above)
Comment (jona-sassenhagen) [reply mne/viz/evoked.py]: Note if we use the MPL defaults as a manual list, that still won't help people who use custom MPL colors.

**Comment by jona-sassenhagen** [comment]
@agramfort 

>what's your bandwidth to finish this?

I'm at a conference, so I'll code during keynotes, my fav coding time :D

**Comment by jona-sassenhagen** [comment]
>having both colors and cmap was indeed also a question for me...

@mmagnuski @agramfort 

`cmap` tells you which colormap to use, `colors` where you are on the colormap.

I'd be open to the idea of enforcing the values of colors to lie between 0 and 1. It has some advantages, some disadvantages. I guess it'd probably better to have it like that.

**Comment by jona-sassenhagen** [comment]
But imagine how it would complicate the examples; we would have to normalise the variable that's color coded, e.g. do a bunch of `(n_letter - min(letters)) / (max(letters) - min(letters))` instead of just `n_letter`.

And it's hard to prepared that list because we retrieve the values from the metadata each time.

File: tutorials/plot_metadata_epochs.py (hunk 14 above)
Comment (jona-sassenhagen) [reply tutorials/plot_metadata_epochs.py]: I've created a variable `letters = epochs.metadata["NumberOfLetters"].unique().astype(str)`, I hope that's sufficient.

**Comment by mmagnuski** [comment]
@jona-sassenhagen 
> But imagine how it would complicate the examples; we would have to normalise the variable that's color coded, e.g. do a bunch of (n_letter - min(letters)) / (max(letters) - min(letters)) instead of just n_letter.

Can't you just do: `np.linspace(0.1, 0.9, num=n_letters)` and then list comprehension?
But for the clarity of the examples, why not just write (for example):
```
colors = {'3': 0.1, '4': 0.3, '5': 0.5}
```
that's more transparent.

**Comment by mmagnuski** [comment]
we could also (at the danger of introducting too much complexity in the input options) normalized color values when they do not fit within 0 - 1 range (or when all are integers).

**Comment by jona-sassenhagen** [comment]
@agramfort @mmagnuski any further input on the possible values of `colors`? I.e., burden the user by enforcing 0-1 range for the dict values.

**Comment by mmagnuski** [comment]
Just as I wrote above: we may consider allowing non-float values and normalize (just like you do now), but also allow to use floats. Using floats can be helpful if one does not want to use the whole colormap space (I remember that sometimes values from the edges of viridis were too bright / too dark to use for lineplots) or wants specific steps (non-linear stepping through colormap). 
I mean - all these situations can be resolved by extracting color values by hand, so I can live without it. :)

Summing up:
* I am ok with enforcing colors to be floats (if one uses colormap)
* I am also ok with allowing both integers and floats
* I am a little less ok, but still can live with allowing only ints

**Comment by jona-sassenhagen** [comment]
> (I remember that sometimes values from the edges of viridis were too bright / too dark to use for lineplots)

That is true, I have the same problem. That's an argument in favour of giving the 0-1 option.

What do you think about:
- if int, linear mapping of ranks
- if float from 0-1, retrieve from cmap

Would that cover your use case?

**Comment by mmagnuski** [comment]
@jona-sassenhagen Yes, that's the second option I am ok with. :)

**Comment by jona-sassenhagen** [comment]
Ok I will go for it, thanks.

**Comment by agramfort** [comment]
without implementing anything how would the code look like for the kiloword
dataset?

**Comment by jona-sassenhagen** [comment]
The current solution would still work. However, there would be another solution where you re-scale the values of `colors`. I.e., something like

```
my_features = epochs.metadata["NumberOfLetters"]
from sklearn.preprocessing import MinMaxScaler
scaler = MinMaxScaler().fit(my_features.reshape(-1, 1))
style_plot["colors"] = {n_letters: scaler.transform(float(n_letters))[0][0]
                        for n_letters in my_features.unique()}
```

for an added 2 or 3 or so lines.

**Comment by mmagnuski** [comment]
or more simply, for example:
```python
colors = {'4': 0.1, '5': 0.3, '6': 0.5, '7': 0.7, '8': 0.9}
```
;)

**Comment by agramfort** [comment]
so it's not a simplification but an extra feature :)

I propose to do this in another PR then.

besides are you guys happy?

**Comment by mmagnuski** [comment]
A separate issue, but not worh another open issue, @jona-sassenhagen are you ok with allowing `ci=False`? I often find myself writing `ci=False` instead of `ci=None`.

**Comment by jona-sassenhagen** [comment]
Ok, I will add a bool option.

**Comment by jona-sassenhagen** [comment]
@agramfort what do you think about @mmagnuski 's suggestion to use, in the plot_metadata_epochs example, using  a manual colors dict, instead of doing it programmatically? It would simplify the code, but would make it less realistic and pythonic.

**Comment by larsoner** [comment]
https://6444-1301584-gh.circle-artifacts.com/0/tmp/circle-artifacts.5ANXZk6/html/auto_tutorials/plot_metadata_epochs.html

**Comment by jona-sassenhagen** [comment]
Oh that looks decent don't you think? I will move the sensor legend to the bottom though.

**Comment by larsoner** [comment]
Yeah that is the only thing I noticed

**Comment by mmagnuski** [comment]
BTW - is the legend a parasite axis (or something like this)? They are much harder to customize (change position etc.), I thinkI had this problem with plot_joint in the past.

**Comment by jona-sassenhagen** [comment]
It's an `mpl_toolkits.axes_grid1.inset_locator.inset_axes` (if you mean the channels), the function is `plot_legend`.

**Comment by jona-sassenhagen** [comment]
Oh nice. We are green.

**Comment by jona-sassenhagen** [comment]
Any chance we can get this merged before tomorrow morning? :)

File: tutorials/plot_metadata_epochs.py [hunk 17]
```diff
@@ -1,4 +1,6 @@
 """
+.. _tut_metadata_epochs:
```
Comment (larsoner) [inline tutorials/plot_metadata_epochs.py]: don't add this, there is a standard ref tag automagically inserted based on filename `sphx_glr_auto_tutorials_plot_metadata_epochs.py`

Comment (jona-sassenhagen) [reply tutorials/plot_metadata_epochs.py]: Ugh, how did this end up there? Must have come with the rebase or something. I didn't do it manually.

File: mne/viz/evoked.py [hunk 18]
```diff
@@ -1705,13 +1757,68 @@ def plot_compare_evokeds(evokeds, picks=list(), gfp=False, colors=None,
...
+            cmap_label = ""
+
+    # dealing with a split legend
+    if split_legend is None:
+        split_legend = cmap is not None  # default to True iff cmap is given
+    if split_legend is True:
+        if colors is None:
+            raise ValueError(
+                "If `split_legend` is True, `colors` must not be None.")
+        import matplotlib.lines as mlines
```
Comment (larsoner) [inline mne/viz/evoked.py]: let's move this to the top of the function so the nested imports / namespace is clear

File: tutorials/plot_visualize_evoked.py [hunk 19]
```diff
@@ -137,6 +137,9 @@
 # with :func:`mne.viz.plot_compare_evokeds`.
 # The plot is styled with dictionary arguments, again using "/"-separated tags.
 # We plot a MEG channel with a strong auditory response.
+#
+# For move advanced plotting using :func:`mne.viz.plot_compare_evokeds`
+# see also :ref:`tut_metadata_epochs`.
```
Comment (jona-sassenhagen) [inline tutorials/plot_visualize_evoked.py]: This too - I didn't put it here.

Should the line be changed? @larsoner

File: mne/viz/evoked.py (hunk 12 above)
Comment (larsoner) [reply mne/viz/evoked.py]: > I would prefer to avoid figure.axes cause we have a varying number of axes.

But based on when it's created it should always be the last axis of the figure. It seems cleaner/more standard to get it as `figure.axes[-1]`.

File: tutorials/plot_visualize_evoked.py (hunk 19 above)
Comment (larsoner) [reply tutorials/plot_visualize_evoked.py]: @agramfort did:

https://github.com/mne-tools/mne-python/pull/4526/commits/cfb5ce6425b9bc368786fa9829bfef35ebad039c

It's fine to leave the new sentences but remove the `_tut_` ref and replace it with the standard one. You can check correctness locally with `cd doc/; make html_dev-noplot`, you should only see 5 (I think?) warnings about gallery images missing in `manual/`.

Comment (jona-sassenhagen) [reply tutorials/plot_visualize_evoked.py]: Oh ok.

**Comment by jona-sassenhagen** [comment]
There, very green.

File: mne/viz/evoked.py (hunk 10 above)
Comment (mmagnuski) [reply mne/viz/evoked.py]: Yes, I was thinking about the current color cycle, not the default one. But I like the default colors, so nevermind for now. :)

**Comment by mmagnuski** [comment]
@jona-sassenhagen I'll take a look this evening, be patient! :)

**Comment by jona-sassenhagen** [comment]
Thanks @mmagnuski !

File: mne/viz/evoked.py [hunk 20]
```diff
@@ -1633,7 +1684,11 @@ def plot_compare_evokeds(evokeds, picks=list(), gfp=False, colors=None,
         ymin, ymax = ylim.get(ch_type, [None, None])
 
     # deal with dict/list of lists and the CI
-    if ci is not None and not (isinstance(ci, np.float) or callable(ci)):
+    if ci is None:
+        ci = False
+    if ci is True:
+        ci = .95
+    elif ci is not False and not (isinstance(ci, np.float) or callable(ci)):
         raise TypeError('ci must be float or callable, got ' + str(type(ci)))
```
Comment (mmagnuski) [inline mne/viz/evoked.py]: nitpick: "ci must be None, bool, float or callable ..."

File: tutorials/plot_metadata_epochs.py [hunk 21]
```diff
@@ -69,36 +70,77 @@
...
+###############################################################################
+# To compare words which are 4, 5, 6, 7 or 8 letters long:
+
+letters = epochs.metadata["NumberOfLetters"].unique().astype(str)
+
+evokeds = dict()
+for n_letters in letters:
+    evokeds[n_letters] = epochs["NumberOfLetters == " + n_letters].average()
+
+style_plot["colors"] = {n_letters: int(float(n_letters))
```
Comment (mmagnuski) [inline tutorials/plot_metadata_epochs.py]: I like how colors are created now, as a side not the `int(float(n_letters))` would not be necessary if NumberOfLetters in metadata was of int type.

**Comment by mmagnuski** [comment]
This already got at least +3 so far so I'll go ahead and merge.
(but I hope for float color option in the future :) - and my nitpicks could be adressed in this future PR if you agree @jona-sassenhagen)

File: tutorials/plot_metadata_epochs.py (hunk 21 above)
Comment (jona-sassenhagen) [reply tutorials/plot_metadata_epochs.py]: Yes, I probably should have done it like that.

File: mne/viz/evoked.py (hunk 20 above)
Comment (jona-sassenhagen) [reply mne/viz/evoked.py]: Addressed, merge if you want.

**Comment by jona-sassenhagen** [comment]
Thanks everyone for reviews!

**Comment by jona-sassenhagen** [comment]
The travis build timed out ... restarted.

**Comment by mmagnuski** [comment]
codecov now complains, but previously all was green. My understanding of codecov is quite limited (I am often confused by why for the same PR one time it is green and another it compains) - so I refer to your intuitions @jona-sassenhagen @larsoner.

**Comment by agramfort** [comment]
@mmagnuski don't merge. I spend a significant of time yesterday to simplify this code.

see https://github.com/mne-tools/mne-python/pull/4755

let me know If I miss some commits there and If none I'll close this PR
